sycl/source/detail/graph_impl.cpp

```diff
@@ -210,21 +226,190 @@ bool graph_impl::clearQueues() {
   return AnyQueuesCleared;
 }
 
-exec_graph_impl::~exec_graph_impl() { MSchedule.clear(); }
+// Check if nodes are empty and if so loop back through predecessors until we
+// find the real dependency.
+void exec_graph_impl::findRealDeps(
+    std::vector<sycl::detail::pi::PiExtSyncPoint> &Deps,
+    std::shared_ptr<node_impl> CurrentNode) {
+  if (CurrentNode->isEmpty()) {
+    for (auto &N : CurrentNode->MPredecessors) {
+      auto NodeImpl = N.lock();
+      findRealDeps(Deps, NodeImpl);
+    }
+  } else {
+    // Verify that the sync point has actually been set for this node.
+    if (auto SyncPoint = MPiSyncPoints.find(CurrentNode);
+        SyncPoint != MPiSyncPoints.end()) {
+      // Check if the dependency has already been added.
+      if (std::find(Deps.begin(), Deps.end(), SyncPoint->second) ==
+          Deps.end()) {
+        Deps.push_back(SyncPoint->second);
+      }
+    } else {
+      assert(false && "No sync point has been set for node dependency.");
+    }
+  }
+}
+
+sycl::detail::pi::PiExtSyncPoint exec_graph_impl::enqueueNodeDirect(
+    sycl::context Ctx, sycl::detail::DeviceImplPtr DeviceImpl,
+    sycl::detail::pi::PiExtCommandBuffer CommandBuffer,
+    std::shared_ptr<node_impl> Node) {
+  std::vector<sycl::detail::pi::PiExtSyncPoint> Deps;
+  for (auto &N : Node->MPredecessors) {
+    findRealDeps(Deps, N.lock());
+  }
+  sycl::detail::pi::PiExtSyncPoint NewSyncPoint;
+  pi_int32 Res = sycl::detail::enqueueImpCommandBufferKernel(
+      Ctx, DeviceImpl, CommandBuffer,
+      *static_cast<sycl::detail::CGExecKernel *>((Node->MCommandGroup.get())),
+      Deps, &NewSyncPoint, nullptr);
+
+  if (Res != pi_result::PI_SUCCESS) {
+    throw sycl::exception(errc::invalid,
+                          "Failed to add kernel to PI command-buffer");
+  }
+
+  return NewSyncPoint;
+}
+
+sycl::detail::pi::PiExtSyncPoint exec_graph_impl::enqueueNode(
+    sycl::context Ctx, std::shared_ptr<sycl::detail::device_impl> DeviceImpl,
+    sycl::detail::pi::PiExtCommandBuffer CommandBuffer,
+    std::shared_ptr<node_impl> Node) {
 
-sycl::event exec_graph_impl::enqueue(
-    const std::shared_ptr<sycl::detail::queue_impl> &Queue) {
-  std::vector<sycl::detail::pi::PiEvent> RawEvents;
+  // Queue which will be used for allocation operations for accessors.
+  auto AllocaQueue = std::make_shared<sycl::detail::queue_impl>(
+      DeviceImpl, sycl::detail::getSyclObjImpl(Ctx), sycl::async_handler{},
+      sycl::property_list{});
+
+  std::vector<sycl::detail::pi::PiExtSyncPoint> Deps;
+  for (auto &N : Node->MPredecessors) {
+    findRealDeps(Deps, N.lock());
+  }
+
+  sycl::detail::EventImplPtr Event =
+      sycl::detail::Scheduler::getInstance().addCG(
+          Node->getCGCopy(), AllocaQueue, CommandBuffer, Deps);
+
+  return Event->getSyncPoint();
+}
+void exec_graph_impl::createURCommandBuffers(sycl::device Device) {
```

Comment:
What is "UR" here?

Reply:
Unified Runtime. We might consider dropping the acronym from the function name to avoid any confusion and connection to the interface that is used under the hood to create command-buffers.

Reply:
I had a feeling it was Unified Runtime. I am not sure I like the impl functions having UR/PI as part of their members. Is there a reason we need to differentiate at this level?

Reply:
Renamed this method to just `createCommandBuffers()`.